### DISCUSSION TITLE
Comparing two floats/doubles: are they the same?

### DISCUSSION THREAD

**Discussion by JacobLotz** [mfem/mfem#3227] [Q&A] [answered]
In nurbs.cpp the class KnotVector has a function `void KnotVector::Difference(const KnotVector &kv, Vector &diff) const` which determines the difference between two knotvectors. The knots of the knotvector are stored in `Vector knot` of which the data consists of doubles.

One of the steps in the process is to loop over the knots in both knotvectors and determine whether the knots are the same. This is done by means of a logical condition `knot(i) == kv[j]` in line 395 of nurbs.cpp. I have come accross an example where the knots where not exactly the same (some kind of round-off(?)) and the function would find a difference that did not exist. Changing the logical condition in line 395 of nurbs.cpp to `(knot(i) - kv[j] < 1e-9)` would solve the issue. 1e-9 was chosen quite randomly but small enough. Knots in the knotvector are usually between 0 and 1 and its order of magnitude depends on the number of knots used. I think that even for very large meshes/knotvectors the order of magnitude of the knots would stay away from 1e-9.

Reading up on this kind of problems on for instance [here (stackoverflow)](https://stackoverflow.com/questions/17333/what-is-the-most-effective-way-for-float-and-double-comparison) indicates that these problems are very delicate and solving it with `(knot(i) - kv[j] < 1e-9)` might cause issues in a later stage. 

I have tried to find a same kind of situation elsewhere in MFEM but could not find one. I think it would make sense to use something like `(knot(i) - kv[j] < 1e-9)` instead of `knot(i) == kv[j]` but someone might have reasoning not to.  Does anyone have advise on how to improve this comparison?

**Comment by vladotomov** [comment, marked as answer]
If the left-hand side is O(1) and is invariant of scaling of the domain, refinements, element size, etc, then it's probably fine.
